src/Juvix/FrontendContextualise/ModuleOpen/Environment.hs

```diff
@@ -100,3 +169,205 @@ lookupModMap s =
+               foldr
+                 (\mod -> Map.insert mod (fmap Implicit qualifiedNames))
+                 map
+                 implicitInner
+           )
+        |> Right
+
+-- this goes off after the checks have passed regarding if the paths
+-- are even possible to resolve
+resolveLoop ::
+  Context.T a b c -> ModuleMap -> Resolve a b c -> Either Error [NameSymbol.T]
+resolveLoop ctx map Res {resolved, notResolved = cantResolveNow} = do
+  map <- foldM addToModMap map fullyQualifyRes
+  --
+  let newResolve = resolveWhatWeCan ctx (qualifyCant map <$> cantResolveNow)
+  --
+  if  | null cantResolveNow ->
+        Right qualifedAns
+      | length (notResolved newResolve) == length cantResolveNow ->
+        Left (CantResolveModules cantResolveNow)
+      | otherwise ->
+        (qualifedAns <>) <$> resolveLoop ctx map newResolve
+  where
+    fullyQualifyRes =
+      Context.resolveName ctx <$> resolved
+    qualifedAns =
+      fmap snd fullyQualifyRes
+    addToModMap map (def, sym) =
+      case def of
+        Context.Record cont _ ->
+          let NameSpace.List {publicL} = NameSpace.toList cont
+           in foldlM
+                ( \map (x, _) ->
+                    case map Map.!? x of
+                      -- this means we already have opened this in another
+                      -- module
+                      Just _ -> Left (AmbiguousSymbol x)
+                      -- if it's already in scope from below or above us
+                      -- don't add to the remapping
+                      Nothing ->
+                        case Context.lookup (pure x) ctx of
+                          Just _ -> Right map
+                          Nothing -> Right (Map.insert x sym map)
+                )
+                map
+                publicL
+        -- should be updated when we can open expressions
+        _ ->
+          Left (UnknownModule sym)
+    qualifyCant newMap term =
+      case newMap Map.!? NameSymbol.hd term of
+        Nothing ->
+          term
+        Just x ->
+          x <> term
+
+-- Since Locals and top level beats opens, we can determine from the
+-- start that any module which tries to open a nested path fails,
+-- yet any previous part succeeds, that an illegal open is happening,
+-- and we can error out immediately
+
+-- | @pathsCanBeResolved@ takes a context and a list of opens,
+-- we then try to resolve if the opens are legal, if so we return
+-- a list of ones that can be determined now, and a list to be resolved
+pathsCanBeResolved ::
+  Context.T a b c -> [NameSymbol.T] -> Either Error (Resolve a b c)
+pathsCanBeResolved ctx opens
+  | fmap firstName (notResolved resFull) == notResolved resFirst =
+    Right resFull
+  | otherwise =
+    Left (OpenNonModule diff)
+  where
+    resFull = resolveWhatWeCan ctx opens
+    resFirst = resolveWhatWeCan ctx (firstName <$> opens)
+    -- O(n log₁₆(n))
+    diff =
+      Set.difference
+        (Set.fromList (notResolved resFull))
```

Comment:
some duplicate computation can be avoided here with `let`

Reply:
yeah will do

Reply:
done